http/dotnet/httpclient/src/Middleware/CompressionHandler.cs

```diff
@@ -0,0 +1,67 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+using System;
+using System.IO.Compression;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Kiota.Http.HttpClient.Middleware
+{
+    /// <summary>
+    /// A <see cref="DelegatingHandler"/> implementation that handles compression.
+    /// </summary>
+    public class CompressionHandler : DelegatingHandler
+    {
+        internal const string GZip = "gzip";
+
+        /// <summary>
+        /// Sends a HTTP request.
+        /// </summary>
+        /// <param name="httpRequest">The <see cref="HttpRequestMessage"/> to be sent.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> for the request.</param>
+        /// <returns></returns>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage httpRequest, CancellationToken cancellationToken)
+        {
+            if(httpRequest == null)
+                throw new ArgumentNullException(nameof(httpRequest));
+
+            StringWithQualityHeaderValue gzipQHeaderValue = new StringWithQualityHeaderValue(GZip);
```

Comment:
shouldn't we compress the request body if we have any? And resend the request uncompressed if we get a 415?

Reply:
According to [this](https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/middleware/CompressionHandler.md#remarks). It looks like we might know that AGS doesn't really support accepting compressed request bodies. Unless things have changed, we would end up with a client SDK sending requests twice for each call.

Reply:
Also captured in https://github.com/microsoft/kiota/issues/598

Reply:
I know they have migrated to aspnet core in the last few months which has had a tendency to fix quirks like these. Unfortunately [after testing](https://gist.github.com/baywet/33b6af6b40fe175324b27b534392ba52) it still seems like it's not decompressing the stream...